### DISCUSSION TITLE
Is it possible to invok a invoke handler from rust code

### DISCUSSION THREAD

**Discussion by chiwa737** [tauri-apps/tauri#6092] [Q&A]
I am adding a small webservice to a exsiting Tauri application as plugin,  so that people can modify the state of tauri managemnet through http request.  But I do not know how to access State from plugin or It could be even better if there is a way to invoke tauri app's invoke_handlers from Rust 

Following code is the implementation of web_service plugin which is added to tauri::app and the web_servce is running well.
```
use http::Uri;
use tauri::{
    plugin::{Builder as PluginBuilder, TauriPlugin},
    Runtime,
};
use actix_web::{get, post, web, App, HttpResponse, HttpServer, Responder};

pub struct Builder {
    port: u16
}

#[actix_web::main]
async fn web_service() -> std::io::Result<()> {
    HttpServer::new(|| {
        App::new()
            .service(web::resource("/").to(|| async { "Web service is running" }))
    })
    .bind(("127.0.0.1", 1234))?
    .run() 
    .await
}

impl Builder {
    pub fn new(port: u16) -> Self {
        Self {
            port
        }
    }
    pub fn build<R: Runtime>(mut self) -> TauriPlugin<R> {
        let port = self.port;
        PluginBuilder::new("localhost")
            .setup(move |app| {
                let asset_resolver = app.asset_resolver();
                std::thread::spawn(move || {
                    web_service()
                });
                Ok(())
            })
            .build()
    }
}



```

**Comment by FabianLars** [comment]
> Is it possible to invok a invoke handler from rust code

Not really. or at least not in the same way as if you call it from the frontend -> you have to inject stuff `AppHandle` and `State` yourself.

> But I do not know how to access State from plugin.

You can use `app` in the setup function to get an AppHandle (`let handle = app.app_handle()`) and then use that handle to get the state, something like this
```rs
use tauri::Manager;
// ...

let state = handle.state::<MyState>()
```

The app handle is `Send + Sync` meaning you can _move_ it around threads as you wish, and it's relatively cheap to clone too.
